### PR TITLE
8331291: java.lang.classfile.Attributes class performs a lot of static initializations

### PR DESCRIPTION
**Pull request by asotona**
Hi,
During performance optimization work on Class-File API as JDK lambda generator we found some static initialization killers.
One of them is `java.lang.classfile.Attributes` with tens of static fields initialized with individual attribute mappers, and common set of all mappers, and static map from attribute names to the mappers.

I propose to turn all the static fields into lazy-initialized static methods and remove `PREDEFINED_ATTRIBUTES` and `standardAttribute(Utf8Entry name)` static mapping method from the `Attributes` API class.
 
Please let me know your comments or objections and please review the [PR](https://github.com/openjdk/jdk/pull/19006) and [CSR](https://bugs.openjdk.org/browse/JDK-8331414), so we can make it into 23.

Thank you,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8331414](https://bugs.openjdk.org/browse/JDK-8331414) to be approved

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8331291: java.lang.classfile.Attributes class performs a lot of static initializations`

### Issues
 * [JDK-8331291](https://bugs.openjdk.org/browse/JDK-8331291): java.lang.classfile.Attributes class performs a lot of static initializations (**Enhancement** - P4)
 * [JDK-8331414](https://bugs.openjdk.org/browse/JDK-8331414): java.lang.classfile.Attributes class performs a lot of static initializations (**CSR**)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author) ⚠️ Review applies to [fd8da774](https://git.openjdk.org/jdk/pull/19006/files/fd8da774565842f710389c9acd243a42afdc4224)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [37f7f63f](https://git.openjdk.org/jdk/pull/19006/files/37f7f63f415ed18c1c9a69048e8c25c490a45bb1)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [dcbaae85](https://git.openjdk.org/jdk/pull/19006/files/dcbaae85b9d3d7484b763178230e6d18429dea55)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19006/head:pull/19006` \
`$ git checkout pull/19006`

Update a local copy of the PR: \
`$ git checkout pull/19006` \
`$ git pull https://git.openjdk.org/jdk.git pull/19006/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19006`

View PR using the GUI difftool: \
`$ git pr show -t 19006`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19006.diff">https://git.openjdk.org/jdk/pull/19006.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19006#issuecomment-2085128751)

### CODE REVIEW

**Comment by liach** [comment]
Nice changes! Remarks
1. The `INSTANCE` fields should be declared `final`, still safe for lazy initialization.
2. Not from this patch, but the `AttributeStability stability()` overrides can be moved to `AbstractAttributeMapper`, stored in a field, to decrease source code size.
3. `AbstractAttributeMapper` might become sealed now that its implementations are no longer anonymous, might move it to be a private nested class of `Attributes` to omit the long list of permits.

**Comment by asotona** [comment]
> Nice changes! Remarks
> 
> 1. The `INSTANCE` fields should be declared `final`, still safe for lazy initialization.
> 2. Not from this patch, but the `AttributeStability stability()` overrides can be moved to `AbstractAttributeMapper`, stored in a field, to decrease source code size.
> 3. `AbstractAttributeMapper` might become sealed now that its implementations are no longer anonymous, might move it to be a private nested class of `Attributes` to omit the long list of permits.

Yes, that is part of the plan for today :)

File: src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java [hunk 1]
```diff
@@ -983,4 +985,88 @@ public byte[] codeArray() {
...
     }
+
+    /**
+     * {@return the attribute mapper for a standard attribute}
+     *
+     * @param name the name of the attribute to find
+     */
+    public static AttributeMapper<?> standardAttribute(Utf8Entry name) {
+        // critical bootstrap path, so no lambdas nor method handles here
+        return switch (name.hashCode()) {
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java:996]: I think we can safely switch over strings, as they are compiled to hashCode switch like what you explicitly have right now. Isn't that the case?

File: test/jdk/jdk/classfile/AttributesTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Testing Attributes API.
```
Comment (liach) [inline test/jdk/jdk/classfile/AttributesTest.java:26]: Can add a line `@bug 8331291`

**Comment by liach** [comment]
Also the `final` modifier addition should be included in the CSR, as it changes the access modifiers even if it has no real impact.

File: src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java (hunk 1 above)
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java:996]: Freshly parsed Utf8Entries conversion to String is expensive and unnecessary. We should be very careful when to ask for the conversion as it significantly affects some benchmarks.

**Comment by asotona** [comment]
> Also the `final` modifier addition should be included in the CSR, as it changes the access modifiers even if it has no real impact.

The class already had only private constructor, so there was no way to extend it.

File: test/jdk/jdk/classfile/AttributesTest.java (hunk 2 above)
Comment (asotona) [reply test/jdk/jdk/classfile/AttributesTest.java:27]: Added, thank you.

File: src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java:996]: You are right, I forgot these are Utf8Entry instead of Strings.

**Comment by liach** [comment]
Yeah, it has no real usage impact but such changes do require CSRs, like https://bugs.openjdk.org/browse/JDK-8305158 for a `final` on `Arrays` class, so you should include the `final` modifier change in your CSR's specdiff.

**Review by liach — APPROVED** [review]
On a side note, will we update JEP 466 to include this patch?

**Review by cl4es — APPROVED** [review]
FWIW code changes looks good to me. There seems to be a number of tests that still need to be updated to use the new methods instead of the old constants.

**Comment by asotona** [comment]
> On a side note, will we update JEP 466 to include this patch?

I hope so, if we get it into 23 ;)

**Comment by asotona** [comment]
> FWIW code changes looks good to me. There seems to be a number of tests that still need to be updated to use the new methods instead of the old constants.

Thank you!
Yes, I'm cleaning the tests right now.

File: src/java.base/share/classes/java/lang/classfile/Attributes.java [hunk 3]
```diff
@@ -24,66 +24,8 @@
...
-import java.lang.classfile.attribute.StackMapTableAttribute;
-import java.lang.classfile.attribute.SyntheticAttribute;
-import java.lang.classfile.constantpool.Utf8Entry;
-import jdk.internal.classfile.impl.AbstractAttributeMapper;
-import jdk.internal.classfile.impl.BoundAttribute;
-import jdk.internal.classfile.impl.CodeImpl;
-import jdk.internal.classfile.impl.AbstractPoolEntry;
-import jdk.internal.classfile.impl.StackMapDecoder;
+import java.lang.classfile.attribute.*;
+import jdk.internal.classfile.impl.AbstractAttributeMapper.*;
```
Comment (vicente-romero-oracle) [inline src/java.base/share/classes/java/lang/classfile/Attributes.java:28]: the second star import is probably unnecessary

Comment (asotona) [reply src/java.base/share/classes/java/lang/classfile/Attributes.java:28]: Thank you for the review!
All the holders/mappers implementations are AbstractAttributeMapper inner classes.

File: src/java.base/share/classes/java/lang/classfile/Attributes.java [hunk 4]
```diff
@@ -207,836 +149,291 @@ public class Attributes {
...
-                    // empty
-                }
-
-                @Override
-                public AttributeMapper.AttributeStability stability() {
-                    return AttributeStability.STATELESS;
-                }
-            };
+    /**
+     * {@return Attribute mapper for the {@code AnnotationDefault} attribute}
```
Comment (liach) [inline src/java.base/share/classes/java/lang/classfile/Attributes.java:153]: Just wondering, can we change `{@code AnnotationDefault}` to `{@value #NAME_ANNOTATION_DEFAULT}`, etc? This way, the names are still rendered as code in Javadoc HTML, but they are generated with links to the constants, and programmers will see these constants and prefer them over hardcoded values.

Comment (asotona) [reply src/java.base/share/classes/java/lang/classfile/Attributes.java:153]: On the other side it is questionable if the attribute names should be exposed in the API. We provide corresponding mappers and attribute models. I don't see a case where user would need to use the attribute names directly.

Comment (liach) [reply src/java.base/share/classes/java/lang/classfile/Attributes.java:153]: Makes sense, we can always add these literals back if we do need them.

**Review by cl4es — APPROVED** [review]
Thank you for this!

**Review by cl4es — APPROVED** [review]
Looks good after revisions.

**Pull request closed without merging**